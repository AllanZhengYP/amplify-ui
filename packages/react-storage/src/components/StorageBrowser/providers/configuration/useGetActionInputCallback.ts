import React from 'react';

import { assertLocationData } from '../../validators';
import { useStore } from '../store';

import { useCredentials } from './credentials';
import { GetActionInput } from './types';
import { LocationData } from '../../actions';
import { assertPrefix, assertScope } from '../../validators/assertLocationData';

export const getErrorMessage = (propertyName: string): string =>
  `Unable to resolve credentials due to invalid value of '${propertyName}'`;

export function useGetActionInputCallback({
  accountId,
  customEndpoint,
  region,
}: {
  accountId?: string;
  customEndpoint?: string;
  region: string;
}): GetActionInput {
  const { getCredentials } = useCredentials();
  const [{ location }] = useStore();
  const { current, key } = location;

  return React.useCallback(
    (location?: LocationData) => {
      // prefer passed in location / prefix over current location in state
      const _location = location ?? current;
      // full S3 prefix is `prefix` in LocationData, but `key` in LocationState
      const _prefix = location?.prefix ?? key;
      assertLocationData(_location, getErrorMessage('locationData'));
      assertPrefix(_prefix, getErrorMessage('prefix'));

      const { bucket, permission, type } = _location;
      // BUCKET/PREFIX grants end with `*`, but object grants do not.
      const scope = `s3://${bucket}/${_prefix}${type === 'OBJECT' ? '' : '*'}`;
      assertScope(scope, getErrorMessage('scope'));

      return {
        accountId,
        bucket,
        credentials: getCredentials({
          scope,
          permission,
        }),
        region,
        customEndpoint,
      };
    },
    [accountId, current, customEndpoint, getCredentials, key, region]
  );
}
