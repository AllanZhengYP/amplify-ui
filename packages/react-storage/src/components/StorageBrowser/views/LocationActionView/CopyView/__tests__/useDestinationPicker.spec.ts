import { act, renderHook, waitFor } from '@testing-library/react';

import * as AmplifyReactCore from '@aws-amplify/ui-react-core';

import * as Store from '../../../../providers/store';
import * as Config from '../../../../providers/configuration';
import {
  DEFAULT_LIST_OPTIONS,
  useDestinationPicker,
} from '../useDestinationPicker';

const mockDispatchStoreAction = jest.fn();
const mockHandleList = jest.fn();
const config = {
  accountId: '123456789012',
  bucket: 'bucket',
  credentials: jest.fn(),
  region: 'us-west-2',
};

const mockItems = [
  {
    key: 'prefix1/',
    lastModified: '2024-11-11T20:36:56.436Z',
    id: 'id',
    size: 10,
    type: 'FOLDER',
  },
  {
    key: 'prefix2/',
    lastModified: '2024-11-11T20:36:56.436Z',
    id: 'id',
    size: 10,
    type: 'FOLDER',
  },
];

describe('useDestinationPicker', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    jest.spyOn(Store, 'useStore').mockReturnValue([
      {
        actionType: 'COPY',
        files: [],
        location: {
          current: {
            prefix: 'test-prefix/',
            bucket: 'bucket',
            id: 'id',
            permission: 'READWRITE',
            type: 'PREFIX',
          },
          path: '',
          key: 'test-prefix/',
        },
        locationItems: {
          fileDataItems: [
            {
              key: 'prefixer/test-file.txt',
              fileKey: 'test-file.txt',
              lastModified: '2024-11-11T20:36:56.436Z' as unknown as Date,
              id: 'id',
              size: 10,
              type: 'FILE',
            },
          ],
        },
      },
      mockDispatchStoreAction,
    ]);

    jest.spyOn(Config, 'useGetActionInput').mockReturnValue(() => config);
  });

  it('should return the correct initial state', async () => {
    jest.spyOn(AmplifyReactCore, 'useDataState').mockReturnValueOnce([
      {
        data: {
          items: mockItems,
          nextToken: 'token',
        },
        hasError: false,
        isLoading: false,
        message: undefined,
      },
      mockHandleList,
    ]);

    const { result } = renderHook(() =>
      useDestinationPicker({
        destinationList: ['prefix1'],
      })
    );

    await waitFor(() => {
      expect(result.current).toMatchSnapshot();
    });
  });

  it('should call handleList once per destinationList change', async () => {
    jest.spyOn(AmplifyReactCore, 'useDataState').mockReturnValue([
      {
        data: {
          items: mockItems,
          nextToken: 'token',
        },
        hasError: false,
        isLoading: false,
        message: undefined,
      },
      mockHandleList,
    ]);

    const { rerender } = renderHook(
      (
        props: { destinationList: string[] } = {
          destinationList: ['prefix1'],
        }
      ) => useDestinationPicker(props)
    );
    await waitFor(() => {
      rerender({
        destinationList: ['prefix1', 'subfolder1'],
      });
      rerender({
        destinationList: ['prefix1'],
      });
      rerender({
        destinationList: ['prefix1'],
      });
      expect(mockHandleList).toHaveBeenCalledTimes(3);
    });
  });

  it('should handle search', () => {
    jest.spyOn(AmplifyReactCore, 'useDataState').mockReturnValueOnce([
      {
        data: {
          items: mockItems,
          nextToken: 'token',
        },
        hasError: false,
        isLoading: false,
        message: undefined,
      },
      mockHandleList,
    ]);
    const { result } = renderHook(() =>
      useDestinationPicker({
        destinationList: ['prefix1'],
      })
    );

    act(() => {
      const state = result.current;
      state.onSearch('moo');
    });

    expect(mockHandleList).toHaveBeenCalledWith({
      config,
      options: {
        ...DEFAULT_LIST_OPTIONS,
        exclude: 'FILE',
        search: { filterKey: 'key', query: 'moo' },
      },
      prefix: 'prefix1/',
    });
  });
});
