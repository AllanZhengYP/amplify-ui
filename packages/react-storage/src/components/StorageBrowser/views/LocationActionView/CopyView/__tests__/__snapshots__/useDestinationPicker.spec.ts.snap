// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDestinationPicker should return the correct initial state 1`] = `
{
  "bucket": "bucket",
  "currentPage": 1,
  "hasError": false,
  "hasNextToken": true,
  "highestPageVisited": 1,
  "isLoading": false,
  "items": [
    {
      "id": "id",
      "key": "prefix1/",
      "lastModified": "2024-11-11T20:36:56.436Z",
      "size": 10,
      "type": "FOLDER",
    },
    {
      "id": "id",
      "key": "prefix2/",
      "lastModified": "2024-11-11T20:36:56.436Z",
      "size": 10,
      "type": "FOLDER",
    },
  ],
  "message": undefined,
  "onPaginate": [Function],
  "onSearch": [Function],
  "pageItems": [
    {
      "id": "id",
      "key": "prefix1/",
      "lastModified": "2024-11-11T20:36:56.436Z",
      "size": 10,
      "type": "FOLDER",
    },
    {
      "id": "id",
      "key": "prefix2/",
      "lastModified": "2024-11-11T20:36:56.436Z",
      "size": 10,
      "type": "FOLDER",
    },
  ],
}
`;
